update-remote-repo.sh

```diff
@@ -30,10 +30,22 @@ sudo chroot chroot-archlinux su user -c 'cd /tmp/qubes-packages-mirror-repo; for
 sudo chroot chroot-archlinux su user -c 'cd /tmp/qubes-packages-mirror-repo; for pkg in `ls -v pkgs/qubes*3.2.*.pkg.tar.xz` ; do repo-add pkgs/qubes-r3.2.db.tar.gz "$pkg";done;'
 
 # Include XEN 4.6.*
-sudo chroot chroot-archlinux su user -c 'cd /tmp/qubes-packages-mirror-repo; for pkg in `ls -v pkgs/qubes-vm-xen-4.6.*.pkg.tar.xz` ; do repo-add pkgs/qubes-r3.1.db.tar.gz "$pkg";done;'
+sudo chroot chroot-archlinux su user -c 'cd /tmp/qubes-packages-mirror-repo; for pkg in `ls -v pkgs/qubes-vm-xen-4.6.*.pkg.tar.xz` ; do repo-add pkgs/qubes-r3.2.db.tar.gz "$pkg";done;'
+
+## release 4.0 Repository
+echo "Generating 4.0 repository"
+
+# Include 3.0 3.1 3.2 and 4.0 packages
+sudo chroot chroot-archlinux su user -c 'cd /tmp/qubes-packages-mirror-repo; for pkg in `ls -v pkgs/qubes*3.0.*.pkg.tar.xz` ; do repo-add pkgs/qubes-r4.0.db.tar.gz "$pkg";done;'
+sudo chroot chroot-archlinux su user -c 'cd /tmp/qubes-packages-mirror-repo; for pkg in `ls -v pkgs/qubes*3.1.*.pkg.tar.xz` ; do repo-add pkgs/qubes-r4.0.db.tar.gz "$pkg";done;'
+sudo chroot chroot-archlinux su user -c 'cd /tmp/qubes-packages-mirror-repo; for pkg in `ls -v pkgs/qubes*3.2.*.pkg.tar.xz` ; do repo-add pkgs/qubes-r4.0.db.tar.gz "$pkg";done;'
```

Comment:
Shouldn't this have different db paths for different versions? If not (i.e. add all packages), why listing specific versions at all?
Also, I recommend using separate qubes-builder instances for different Qubes versions to avoid linking with wrong libraries (installed in build environment). And relying on package version to tell for which Qubes version it is can be fragile - for example all app-* are the same regardless of Qubes version.

Reply:
Actually, I did not included any app until now and I was destroying the chroot before building for another version. That was making things much more simple.

But as you suggest, it can be a problem for apps. Are apps relying on qubes libraries ?
How do you deal with multiple qubes-builders ? Are you sharing caches between builder instances ?

Reply:
> Are apps relying on qubes libraries ?

In theory could be. For example linked to libvchan, or libqrexec-utils. In practice none of them are currently. But for example dbus component (used by desktop-linux-manager - new widgets) have version 1.x. right now. And thanks to Admin API, can also be used from a VM (in Qubes 4.1 will be - from GUI VM).

> How do you deal with multiple qubes-builders ? Are you sharing caches between builder instances ?

I use completely separate directories, like `builder-r3.2`, `builder-r4.0`, without sharing any files between them. Require some more disk space and network bandwidth, but this way I can be sure packages use the right build environments. In some cases, even in different VMs. See [here](https://github.com/QubesOS/qubes-infrastructure/)

Reply:
I will try something similar, but I'm limited both in space and bandwith as I'm using a single system for everything. Which is feasible thanks to Qubes (except for Gaming and some Virtual Machines based on other hypervisors).
Can you still accept the Pull Request for now, if it is the only comment ? I need to think about how to switch pacman repositories directly when updating packages (which is now .disabled by default). And I will probably purchase a storage extension by the way...

Reply:
Related to the qubes apps versioning, is it possible to follow Qubes versioning as soon as there are dependencies on a specific Qubes version, while keeping version 1.X for all other apps ?

Reply:
Hmm, so how do you separate building non-apps packages for different Qubes versions? For example Qubes 4.0 use different Xen version than 3.2, so vchan etc should be linked with different libraries.

Reply:
I destroy the chroot-archlinux and delete built pkgs while keeping the caches. Then I build each Qube version starting from the oldest.
Of course there are a lot of room for errors (now that you talk about it, if I don't cleanup the packages properly, Qubes4.0 packages would be used in priority in the chroot as their version is higher).
Using a separate builder for each version would definitively limit such errors.

Reply:
If you want to share cache (which should be safe), you can symlink one to another.

Reply:
BTW does Archlinux have native tool to build packages, _including build env setup_? Like `mock` on Fedora, or `pbuilder` on Debian.

Reply:
What do you mean by *build env setup* ?

In archlinux, makepkg is used to build packages and the current environement is used as the build environment. The tool fakeroot is used to isolate things at build time for safety reasons (avoid overwriting the system).

I think that as archlinux is a rolling release, a build environment makes little sense.

Reply:
This means you have to install all required make-dependencies to build things.
There are pacman wrappers such as `yaourt` that deal with everything for you including make-dependencies.
There are also some non native projects to use chroot build environment and create repositories automatically. Personally, I use `arch-ppa`.

Reply:
But for example you need to install all build dependencies, make sure versions are appropriate etc. Additionally, doing that in chroot is a bonus (so you can isolate build env, from your normal system - have installed _only_ what is needed; it helps with reproducible builds).

> There are also some non native projects to use chroot build environment and create repositories automatically. Personally, I use arch-ppa.

Exactly the think I was asking for, thanks :)

Reply:
In archlinux the concept of versionning is limited. They try to always use the last versions and there are little exceptions.
Installing dependencies is very fast compared to dnf and there are commands to find unused dependencies to clean up the system.
Finally the chroot part is *somehow* done by fakeroot to at least avoid overwritting system files.
For this reason, most people just upload their PKGBUILD on AUR (community package repository), and use pacman wrappers such as yaourt that makes straightforward to build and install all packages and dependencies in /tmp and cleanup the system after that just by running `yaourt -S qubes-vm-core`.
Reproducible builds will be quite hard to achieve in archlinux as you highly depends on the freshness of your package mirror, and old package versions are often removed from mirrors as soon as new ones are available.